[JSC][GLib] Provide deferred promise creation api

tmaillart opened the pull request:
#### f4cbe7bf3972971aa4feba14a2e466b81af92980
<pre>
[JSC][GLib] Provide deferred promise creation api
<a href="https://bugs.webkit.org/show_bug.cgi?id=283708">https://bugs.webkit.org/show_bug.cgi?id=283708</a>

Reviewed by Carlos Garcia Campos.

This commit implements an API to create promises in JSC GLib.
It is a copy from the Objective-C valueWithNewPromiseInContext.
An executor is required, it will be invoked during promise initialization
and it receives the @resolve and @reject objects than can be called to resolve
or reject the promise. It is called like a js function, so exceptions raised
during the executor invocation will not be propagated to the context, but
handled by the promise causing a rejection.

* Source/JavaScriptCore/API/glib/JSCValue.cpp:
(jsc_value_new_promise):
* Source/JavaScriptCore/API/glib/JSCValue.h.in:
* Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
(fooSuccessSource):
(fooSuccessExecutor):
(getAsyncResolvedPromise):
(fooFailureSource):
(fooFailureExecutor):
(getAsyncRejectedPromise):
(getRejectedPromise):
(getExceptionPromise):
(getResolvedPromise):
(testJSCPromises):

Canonical link: <a href="https://commits.webkit.org/288688@main">https://commits.webkit.org/288688@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c47c978d72f30224b2d7a04c8b5b39ce8b7cf171

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84023 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3641 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38324 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89098 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35031 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86108 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3732 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11609 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65349 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23192 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87069 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2782 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76343 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45642 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2713 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30567 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34080 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/76987 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73676 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31325 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90476 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/83040 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11288 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8174 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73803 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11512 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72172 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73020 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17310 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/2634 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13018 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11240 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16712 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/105459 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11088 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25478 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14564 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12860 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/JavaScriptCore/API/glib/JSCValue.cpp:
```diff
@@ -2145,3 +2145,30 @@ char* jsc_value_to_json(JSCValue* value, unsigned indent)
+/**
+ * jsc_value_new_deferred_promise:
+ * @context: a #JSCContext
+ * @value: a #JSCValue
```
The list of parameters is missing `resolve` and `reject`, which need to be marked `(out) (transfer full) (nullable)` to indicate that they are new values returned by the function, that the caller takes their ownership, and that it is acceptable to pass `NULL`:

```suggestion
 * @value: a #JSCValue
 * @resolve: (out) (transfer full) (nullable): a callable #JSCValue.
 * @reject: (out) (transfer full) (nullable): a callable #JSCValue.
```

aperezdc commented on Source/JavaScriptCore/API/glib/JSCValue.cpp:
```diff
@@ -2145,3 +2145,30 @@ char* jsc_value_to_json(JSCValue* value, unsigned indent)
+ *
+ * Create a new Promise returning handles to resolve and reject method
+ *
+ * Returns: (transfer full): a defered promise
```
Small typo, and it would be good to indicate the returned value is an object:

```suggestion
 * Returns: (transfer full): a deferred promise object
```

Adding a small code example to the documentation about how to use a promise object created with this API would be nice, too.

tmaillart replied:
I agree, by documentation you mean here in the comment section or you have a specific file in mind ?

aperezdc commented on Source/JavaScriptCore/API/glib/JSCValue.cpp:
```diff
@@ -2145,3 +2145,30 @@ char* jsc_value_to_json(JSCValue* value, unsigned indent)
+    JSObjectRef resolveObj, rejectObj;
+    JSCValue* ret;
+
+    ret = jscContextGetOrCreateValue(context, JSObjectMakeDeferredPromise(jsContext, &resolveObj, &rejectObj, nullptr)).leakRef();
```
We want to pass `&resolveObj` and `&rejectObj` only if `resolve` and `reject` are non-null, respectively, to avoid having to destroy them immediately after creation in case the caller will not be using them. Also, the call may be all in one line:

```suggestion
    JSObjectRef resolveObj = nullptr;
    JSObjectRef rejectObj = nullptr;
    JSCValue* ret = jscContextGetOrCreateValue(context, JSObjectMakeDeferredPromise(jsContext, resolve ? &resolveObj : nullptr, reject ? &rejectObj : nullptr, nullptr)).leakRef();
```

aperezdc replied:
@carlosgcampos This is ignoring the `exception` argument to `JSObjectMakeDeferredPromise()`, do you know if that would be okay here?

carlosgcampos replied:
Yes, exception should be handled

aperezdc reviewed (commented):
I left a few comments, and given this adds new API it would be better for a second pair of eyes to take a look as well. Maybe @carlosgcampos and/or @mcatanzaro can do a second review here.

The new API function name sounds fine to me, and I won't oppose. I do wonder if we could do with the shorter `jsc_value_new_promise()`. I don't consider myself an expert writing JS code, and yet to me it is quite clear that a promise is always deferred right after creation without needing to make it explicit in the function name.

mcatanzaro commented on Source/JavaScriptCore/API/glib/JSCValue.cpp:
```diff
@@ -2145,3 +2145,31 @@ char* jsc_value_to_json(JSCValue* value, unsigned indent)
+/**
+ * jsc_value_new_deferred_promise:
+ * @context: a #JSCContext
+ * @value: a #JSCValue
```
This extra @value parameter here is why the build is failing.

mcatanzaro reviewed (changes requested):
Can you please add a test in Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp? That will also demonstrate for us reviewers precisely how the new API is expected to be used.

We might also want to add a code example to the documentation comment.

carlosgcampos reviewed (changes requested):
Thanks for the patch. all patches including new API should include API tests. I think we might want to handle this similar to the Objc API, the new function receives a callback where resolve and reject are received, and if an exception is generated in the callback it causes the promise to be rejected without generating an exception the context.

carlosgcampos commented on Source/JavaScriptCore/API/glib/JSCValue.cpp:
```diff
@@ -2145,3 +2145,33 @@ char* jsc_value_to_json(JSCValue* value, unsigned indent)
+        *resolve = jscContextGetOrCreateValue(context, resolveObj).leakRef();
+    if (reject)
+        *reject = jscContextGetOrCreateValue(context, rejectObj).leakRef();
+    return ret;
```
What about promise rejected by an exception? in case of exception the promise is rejected and the exception can be handled in catch(), and not exposed to the context. That's why I suggested to use callback that receives resolve/reject. That way we can push/pop a function call context to catch exceptions. See the [Objc API](https://developer.apple.com/documentation/javascriptcore/jsvalue/init(newpromisein:fromexecutor:)?changes=lat____4&language=objc)

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3963,6 +4001,23 @@ static void testJSCPromises()
+        GRefPtr<JSCValue> result = adoptGRef(jsc_context_evaluate(context.get(), "result = -1; f = new Fooo(); f.getPromise().then(function (value) { result = 1; }, function (error) { result = 0; });", -1));
+        g_main_loop_run(m);
+        result = adoptGRef(jsc_context_get_value(context.get(), "result"));
+        g_assert_true(jsc_value_to_int32(result.get()) == 1);
```
We need more test cases:
* Exception raised when creating the promise.
* Invoke then/catch methods from the API on the promise object.
* Resolve and rejection on promise creation and later
* Reject by raising an exception (both by setting the exception from the api and js)

carlosgcampos commented on Source/JavaScriptCore/API/glib/JSCValue.cpp:
```diff
@@ -2145,3 +2145,43 @@ char* jsc_value_to_json(JSCValue* value, unsigned indent)
+ * @executor: (scope call) (nullable) (closure user_data): an initialization callback
+ * @user_data: (nullable): user data passed in @executor
+ *
+ * Create a new Promise returning handles to resolve and reject method
```
This is not accurate now, we should document here that executor will be invoked during promise initialization and it receives the resolve and reject objects than can be called to resolve or reject the promise. We should also document that the executor is called like a js function, so exceptions raised during the executor invocation will  not be propagated to the context, but handled by the promise causing a rejection.

carlosgcampos commented on Source/JavaScriptCore/API/glib/JSCValue.cpp:
```diff
@@ -2145,3 +2145,43 @@ char* jsc_value_to_json(JSCValue* value, unsigned indent)
+    JSObjectRef resolveObj = nullptr;
+    JSObjectRef rejectObj = nullptr;
+    JSValueRef exception = nullptr;
+    JSValueRef promise = JSObjectMakeDeferredPromise(jsContext, &resolveObj, &rejectObj, &exception);
```
```suggestion
    JSObjectRef promise = JSObjectMakeDeferredPromise(jsContext, &resolveObj, &rejectObj, &exception);
```

carlosgcampos commented on Source/JavaScriptCore/API/glib/JSCValue.cpp:
```diff
@@ -2145,3 +2145,43 @@ char* jsc_value_to_json(JSCValue* value, unsigned indent)
+    GRefPtr<JSCValue> reject = jscContextGetOrCreateValue(context, rejectObj);
+    executor(resolve.get(), reject.get(), userData);
+    JSCException* unhandledException = jsc_context_get_exception(context);
+    if (unhandledException)
```
```suggestion
    if (JSCException* unhandledException = jsc_context_get_exception(context))
```

carlosgcampos commented on Source/JavaScriptCore/API/glib/JSCValue.h.in:
```diff
@@ -322,6 +322,15 @@ JSC_API char *
 
+typedef void  (*JSCExecutor)              (JSCValue             *resolve,
+                                           JSCValue             *reject,
+                                           gpointer             user_data);
```
This also need to be documented

mcatanzaro replied:
```suggestion
                                           JSCValue             *reject,
                                           gpointer              user_data);
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3905,6 +3905,112 @@ static void testJSCExceptions()
+    JSCValue* resolve;
+    JSCValue* reject;
+    JSCContext* context;
+    GMainLoop* m;
```
```suggestion
    GMainLoop* mainLoop;
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3963,6 +4069,70 @@ static void testJSCPromises()
+        jsc_class_add_method(foooClass, "getAsyncRejectedPromise", G_CALLBACK(getAsyncRejectedPromise), nullptr, nullptr, JSC_TYPE_VALUE, 0, G_TYPE_NONE);
+        jsc_class_add_method(foooClass, "getResolvedPromise", G_CALLBACK(getResolvedPromise), nullptr, nullptr, JSC_TYPE_VALUE, 0, G_TYPE_NONE);
+        jsc_class_add_method(foooClass, "getRejectedPromise", G_CALLBACK(getRejectedPromise), nullptr, nullptr, JSC_TYPE_VALUE, 0, G_TYPE_NONE);
+        jsc_class_add_method(foooClass, "getExceptionPromise", G_CALLBACK(getExceptionPromise), nullptr, nullptr, JSC_TYPE_VALUE, 0, G_TYPE_NONE);
```
Why do we need a new class for this test? and why subtests are actually methods of the class?

tmaillart replied:
There is no particular reason, I wrote the tests based on what we did for our custom js api

carlosgcampos replied:
I find it more difficult to follow, to be honest.

tmaillart replied:
Ok I removed the class

carlosgcampos reviewed (changes requested):
The API now looks great, thanks! We just need to improve a bit documentation and tests.

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3905,6 +3905,87 @@ static void testJSCExceptions()
     }
 }
 
+struct foo {
```
```suggestion
struct Foo {
```

carlosgcampos replied:
But maybe we can use a better name than Foo... Something like AsyncPromiseTestData?

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3905,6 +3905,87 @@ static void testJSCExceptions()
+
+static gboolean foooSuccessSource(gpointer data)
+{
+    struct foo* foo = (struct foo*) data;
```
Don't use C-style casts, use C++ instead.

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3905,6 +3905,87 @@ static void testJSCExceptions()
 
+struct foo {
+    GMainLoop* mainLoop;
+    gpointer data;
```
```suggestion
    GRefPtr<JSCValue> value;
```

carlosgcampos replied:
data is always a JSCValue, right?

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3905,6 +3905,87 @@ static void testJSCExceptions()
 }
 
+struct foo {
+    GMainLoop* mainLoop;
```
```suggestion
    GRefPtr<GMainLoop> mainLoop;
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3905,6 +3905,87 @@ static void testJSCExceptions()
+    return FALSE;
+}
+
+static void foooSuccessExecutor(JSCValue* resolve, JSCValue* reject, gpointer user_data)
```
```suggestion
static void foooSuccessExecutor(JSCValue* resolve, JSCValue* reject, gpointer userData)
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3905,6 +3905,87 @@ static void testJSCExceptions()
+    gpointer data;
+};
+
+static gboolean foooSuccessSource(gpointer data)
```
```suggestion
static gboolean foooSuccessSource(gpointer userData)
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3905,6 +3905,87 @@ static void testJSCExceptions()
+{
+    struct foo* foo = (struct foo*) data;
+    g_object_unref(jsc_value_function_call(static_cast<JSCValue*>(foo->data), G_TYPE_NONE));
+    g_clear_pointer(&foo->data, g_object_unref);
```
Using GRefPtr you only need:

```
foo->data = nullptr;
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3905,6 +3905,87 @@ static void testJSCExceptions()
+    foo->data = g_object_ref(resolve);
+}
+
+static JSCValue* getAsyncResolvedPromise(struct foo* foo)
```
```suggestion
static JSCValue* getAsyncResolvedPromise(Foo* foo)
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3905,6 +3905,87 @@ static void testJSCExceptions()
+static JSCValue* getAsyncResolvedPromise(struct foo* foo)
+{
+    JSCValue* promise = jsc_value_new_promise(jsc_context_get_current(), foooSuccessExecutor, foo);
+    g_assert_true(promise);
```
```suggestion
    g_assert_nonnull(promise);
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3905,6 +3905,87 @@ static void testJSCExceptions()
+    return promise;
+}
+
+static gboolean foooFailureSource(gpointer data)
```
```suggestion
static gboolean foooFailureSource(gpointer userData)
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3905,6 +3905,87 @@ static void testJSCExceptions()
+    foo->data = g_object_ref(reject);
+}
+
+static JSCValue* getAsyncRejectedPromise(struct foo* foo)
```
```suggestion
static JSCValue* getAsyncRejectedPromise(Foo* foo)
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3905,6 +3905,87 @@ static void testJSCExceptions()
+static JSCValue* getAsyncRejectedPromise(struct foo* foo)
+{
+    JSCValue* promise = jsc_value_new_promise(jsc_context_get_current(), foooFailureExecutor, foo);
+    g_assert_true(promise);
```
```suggestion
    g_assert_nonnull(promise);
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3905,6 +3905,87 @@ static void testJSCExceptions()
+    jsc_context_throw(jsc_context_get_current(), "exception in executor");
+}
+
+static JSCValue* getRejectedPromise(struct foo* foo)
```
```suggestion
static JSCValue* getRejectedPromise(Foo* foo)
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3905,6 +3905,87 @@ static void testJSCExceptions()
+    return promise;
+}
+
+static void foooRejectedExecutor(JSCValue* resolve, JSCValue* reject, gpointer user_data)
```
```suggestion
static void foooRejectedExecutor(JSCValue* resolve, JSCValue* reject, gpointer userData)
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3905,6 +3905,87 @@ static void testJSCExceptions()
+    g_object_unref(jsc_value_function_call(reject, G_TYPE_BOOLEAN, TRUE, G_TYPE_NONE));
+}
+
+static void foooExceptionExecutor(JSCValue* resolve, JSCValue* reject, gpointer user_data)
```
```suggestion
static void foooExceptionExecutor(JSCValue* resolve, JSCValue* reject, gpointer userData)
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3905,6 +3905,87 @@ static void testJSCExceptions()
+
+static JSCValue* getRejectedPromise(struct foo* foo)
+{
+    return jsc_value_new_promise(jsc_context_get_current(), foooRejectedExecutor, nullptr);
```
The executor could be a lambda.

```
return jsc_value_new_promise(jsc_context_get_current(), +[](JSCValue*, JASCValue* reject, gpointer) {
    g_object_unref(jsc_value_function_call(reject, G_TYPE_BOOLEAN, TRUE, G_TYPE_NONE));
}, nullptr);
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3905,6 +3905,87 @@ static void testJSCExceptions()
+    return jsc_value_new_promise(jsc_context_get_current(), foooRejectedExecutor, nullptr);
+}
+
+static JSCValue* getExceptionPromise(struct foo* foo)
```
```suggestion
static JSCValue* getExceptionPromise(Foo* foo)
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3905,6 +3905,87 @@ static void testJSCExceptions()
+
+static JSCValue* getExceptionPromise(struct foo* foo)
+{
+    return jsc_value_new_promise(jsc_context_get_current(), foooExceptionExecutor, nullptr);
```
Same here, the executor could be a lambda

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3905,6 +3905,87 @@ static void testJSCExceptions()
+static JSCValue* getResolvedPromise(struct foo* foo)
+{
+    return jsc_value_new_promise(jsc_context_get_current(), foooResolvedExecutor, nullptr);
+}
```
And same comments here

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3963,6 +4044,73 @@ static void testJSCPromises()
+
+    {
+        GMainLoop* m = g_main_loop_new(NULL, FALSE);
+        struct foo foo = { .mainLoop = m, .data = nullptr };
```
Foo foo = { adoptGRef(g_main_loop_new(nullptr, FALSE)), nullptr };

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3963,6 +4044,73 @@ static void testJSCPromises()
+        GRefPtr<JSCContext> context = adoptGRef(jsc_context_new());
+
+        GRefPtr<JSCValue> function;
+        function = adoptGRef(jsc_value_new_function(context.get(), "getAsyncResolvedPromise", G_CALLBACK(getAsyncResolvedPromise), &foo, nullptr, JSC_TYPE_VALUE, 0, G_TYPE_NONE));
```
```suggestion
        GRefPtr<JSCValue> function = adoptGRef(jsc_value_new_function(context.get(), "getAsyncResolvedPromise", G_CALLBACK(getAsyncResolvedPromise), &foo, nullptr, JSC_TYPE_VALUE, 0, G_TYPE_NONE));
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3963,6 +4044,73 @@ static void testJSCPromises()
+        function = adoptGRef(jsc_value_new_function(context.get(), "getAsyncResolvedPromise", G_CALLBACK(getAsyncResolvedPromise), &foo, nullptr, JSC_TYPE_VALUE, 0, G_TYPE_NONE));
+        jsc_context_set_value(context.get(), "getAsyncResolvedPromise", function.get());
+        GRefPtr<JSCValue> result;
+        result = adoptGRef(jsc_context_evaluate(context.get(), "result = -1; getAsyncResolvedPromise().then(function (value) { result = 1; }, function (error) { result = 2; });", -1));
```
```suggestion
        GRefPtr<JSCValue> result = adoptGRef(jsc_context_evaluate(context.get(), "result = -1; getAsyncResolvedPromise().then(function (value) { result = 1; }, function (error) { result = 2; });", -1));
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3963,6 +4044,73 @@ static void testJSCPromises()
+        jsc_context_set_value(context.get(), "getAsyncResolvedPromise", function.get());
+        GRefPtr<JSCValue> result;
+        result = adoptGRef(jsc_context_evaluate(context.get(), "result = -1; getAsyncResolvedPromise().then(function (value) { result = 1; }, function (error) { result = 2; });", -1));
+        g_main_loop_run(m);
```
```suggestion
        g_main_loop_run(foo->mainLoop.get());
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3963,6 +4044,73 @@ static void testJSCPromises()
+        result = adoptGRef(jsc_context_evaluate(context.get(), "result = -1; getAsyncResolvedPromise().then(function (value) { result = 1; }, function (error) { result = 2; });", -1));
+        g_main_loop_run(m);
+        result = adoptGRef(jsc_context_get_value(context.get(), "result"));
+        g_assert_true(jsc_value_to_int32(result.get()) == 1);
```
```suggestion
        g_assert_cmpint(jsc_value_to_int32(result.get()), ==, 1);
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3963,6 +4044,73 @@ static void testJSCPromises()
+        result = adoptGRef(jsc_context_evaluate(context.get(), "result = -1; getAsyncRejectedPromise().then(function (value) { result = 1; }, function (error) { result = 2; });", -1));
+        g_main_loop_run(m);
+        result = adoptGRef(jsc_context_get_value(context.get(), "result"));
+        g_assert_true(jsc_value_to_int32(result.get()) == 2);
```
```suggestion
        g_assert_cmpint(jsc_value_to_int32(result.get()), ==, 2);
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3963,6 +4044,73 @@ static void testJSCPromises()
+        jsc_context_set_value(context.get(), "getResolvedPromise", function.get());
+        adoptGRef(jsc_context_evaluate(context.get(), "result = -1; getResolvedPromise().then(function (value) { result = 1; }, function (error) { result = 2; });", -1));
+        result = adoptGRef(jsc_context_get_value(context.get(), "result"));
+        g_assert_true(jsc_value_to_int32(result.get()) == 1);
```
Use cmpint always when comparing ints, please

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3963,6 +4044,73 @@ static void testJSCPromises()
+        GRefPtr<JSCValue> promise, resolve_func, reject_func;
+        promise = adoptGRef(jsc_context_evaluate(context.get(), "result = -1; p = getResolvedPromise();", -1));
+        resolve_func = adoptGRef(jsc_context_evaluate(context.get(), "res = function (a) { result = 1; }", -1));
+        reject_func  = adoptGRef(jsc_context_evaluate(context.get(), "rej = function (a) { result = 2; }", -1));
```
```suggestion
        GRefPtr<JSCValue> promise = adoptGRef(jsc_context_evaluate(context.get(), "result = -1; p = getResolvedPromise();", -1));
        GRefPtr<JSCValue> resolveFunc = adoptGRef(jsc_context_evaluate(context.get(), "res = function (a) { result = 1; }", -1));
        GRefPtr<JSCValue> rejectFunc  = adoptGRef(jsc_context_evaluate(context.get(), "rej = function (a) { result = 2; }", -1));
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3963,6 +4035,66 @@ static void testJSCPromises()
+
+        GRefPtr<JSCValue> promise = adoptGRef(jsc_context_evaluate(context.get(), "result = -1; p = getResolvedPromise();", -1));
+        GRefPtr<JSCValue> resolve_func = adoptGRef(jsc_context_evaluate(context.get(), "res = function (a) { result = 1; }", -1));
+        GRefPtr<JSCValue> reject_func  = adoptGRef(jsc_context_evaluate(context.get(), "rej = function (a) { result = 2; }", -1));
```
```suggestion
        GRefPtr<JSCValue> resolveFunc = adoptGRef(jsc_context_evaluate(context.get(), "res = function (a) { result = 1; }", -1));
        GRefPtr<JSCValue> rejectFunc  = adoptGRef(jsc_context_evaluate(context.get(), "rej = function (a) { result = 2; }", -1));
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3963,6 +4035,66 @@ static void testJSCPromises()
     }
+
+    {
+        AsyncPromiseTestData foo = { .mainLoop = adoptGRef(g_main_loop_new(NULL, FALSE)), .value = nullptr };
```
```suggestion
        AsyncPromiseTestData foo = { .mainLoop = adoptGRef(g_main_loop_new(nullptr, FALSE)), .value = nullptr };
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3905,6 +3905,78 @@ static void testJSCExceptions()
+    return FALSE;
+}
+
+static void foooFailureExecutor(JSCValue* resolve, JSCValue* reject, gpointer userData)
```
```suggestion
static void foooFailureExecutor(JSCValue*, JSCValue* reject, gpointer userData)
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3905,6 +3905,78 @@ static void testJSCExceptions()
+    return FALSE;
+}
+
+static void foooSuccessExecutor(JSCValue* resolve, JSCValue* reject, gpointer userData)
```
```suggestion
static void foooSuccessExecutor(JSCValue* resolve, JSCValue*, gpointer userData)
```

carlosgcampos reviewed (changes requested):
Thanks! The commit message is still wrong, though

carlosgcampos reviewed (changes requested):
The commit message is still wrong. It should be:

```
[JSC][GLib] Provide deferred promise creation API
https://bugs.webkit.org/show_bug.cgi?id=283708

Reviewed by NOBODY (OOPS!).

Explanation of what this commit is doing

List of files
```

In your case:

1. The bug title is missing
2. The line "Explanation of why this fixes the bug (OOPS!)." needs to be replaced with the actual explanation fo what this patch is doing
3. The list of files doesn't correspond to this patch.

The patch is ok now, you just need to fix the commit message. Thanks!

mcatanzaro commented on Source/JavaScriptCore/API/glib/JSCValue.cpp:
```diff
@@ -2145,3 +2145,61 @@ char* jsc_value_to_json(JSCValue* value, unsigned indent)
+ *
+ * It is called like a js function, so exceptions raised will not be propagated
+ * to the context, but handled by the promise causing a rejection.
+ * @resolve and @reject can be reffed for later use to handle async tasks completion.
```
The doc comments need more work. This isn't sufficient to explain what a JSCExecutor is or why you would want to use one. I see it's really simple -- just a function that you pass to `jsc_value_new_promise()` -- so might as well say that here.

mcatanzaro replied:
```suggestion
 * @resolve and @reject can be reffed for later use to handle async task completion.
```

Can you please add a code example to demonstrate how this works? I think currently you'd need to look at TestJSC.cpp to figure out how it's supposed to be used.

mcatanzaro commented on Source/JavaScriptCore/API/glib/JSCValue.cpp:
```diff
@@ -2145,3 +2145,61 @@ char* jsc_value_to_json(JSCValue* value, unsigned indent)
+ *
+ * Function invoked at promise initialization
+ *
+ * It is called like a js function, so exceptions raised will not be propagated
```
```suggestion
 * It is called like a JS function, so exceptions raised will not be propagated
```

Or better:

```suggestion
 * It is called like a JavaScript function, so exceptions raised will not be propagated
```

mcatanzaro commented on Source/JavaScriptCore/API/glib/JSCValue.cpp:
```diff
@@ -2145,3 +2145,61 @@ char* jsc_value_to_json(JSCValue* value, unsigned indent)
+ * @executor: (scope call) (closure user_data): an initialization callback
+ * @user_data: (nullable): user data passed in @executor
+ *
+ * Create a new Promise, executor will be invoked during promise initialization
```
This is a comma splice:

```suggestion
 * Creates a new Promise. @executor will be invoked during promise initialization
```

mcatanzaro commented on Source/JavaScriptCore/API/glib/JSCValue.cpp:
```diff
@@ -2145,3 +2145,61 @@ char* jsc_value_to_json(JSCValue* value, unsigned indent)
+ *
+ * Create a new Promise, executor will be invoked during promise initialization
+ * and it receives the @resolve and @reject objects than can be called to resolve
+ * or reject the promise. It is called like a js function, so exceptions raised
```
```suggestion
 * or reject the promise. It is called like a JavaScript function, so exceptions raised
```

mcatanzaro commented on Source/JavaScriptCore/API/glib/JSCValue.h.in:
```diff
@@ -322,6 +322,15 @@ JSC_API char *
+JSC_API JSCValue *
+jsc_value_new_promise                     (JSCContext           *context,
+                                           JSCExecutor          executor,
+                                           gpointer             user_data);
```
```suggestion
jsc_value_new_promise                     (JSCContext           *context,
                                           JSCExecutor           executor,
                                           gpointer              user_data);
```

mcatanzaro commented on Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
```diff
@@ -3905,6 +3905,78 @@ static void testJSCExceptions()
+    GRefPtr<JSCValue> value;
+};
+
+static gboolean foooSuccessSource(gpointer userData)
```
Why fooo instead of foo?

carlosgcampos commented on Source/JavaScriptCore/API/glib/JSCValue.cpp:
```diff
@@ -2145,3 +2145,85 @@ char* jsc_value_to_json(JSCValue* value, unsigned indent)
+ * {
+ *     return jsc_value_new_promise(jsc_context_get_current(), fooExecutor, foo);
+ * }
+ * ]|
```
I'm not sure this example is very clear to be honest, since it's copied from the unit tests without context about what Foo is. I don't want to block this more, because the implementation and unit tests are ok. So, let's remove the code example for now, we can improve the docs later in a follow up.

tmaillart committed:
[JSC][GLib] Provide deferred promise creation api
https://bugs.webkit.org/show_bug.cgi?id=283708

Reviewed by Carlos Garcia Campos.

This commit implements an API to create promises in JSC GLib.
It is a copy from the Objective-C valueWithNewPromiseInContext.
An executor is required, it will be invoked during promise initialization
and it receives the @resolve and @reject objects than can be called to resolve
or reject the promise. It is called like a js function, so exceptions raised
during the executor invocation will not be propagated to the context, but
handled by the promise causing a rejection.

* Source/JavaScriptCore/API/glib/JSCValue.cpp:
(jsc_value_new_promise):
* Source/JavaScriptCore/API/glib/JSCValue.h.in:
* Tools/TestWebKitAPI/Tests/JavaScriptCore/glib/TestJSC.cpp:
(fooSuccessSource):
(fooSuccessExecutor):
(getAsyncResolvedPromise):
(fooFailureSource):
(fooFailureExecutor):
(getAsyncRejectedPromise):
(getRejectedPromise):
(getExceptionPromise):
(getResolvedPromise):
(testJSCPromises):

Canonical link: https://commits.webkit.org/288688@main

ntrrgc commented on Source/JavaScriptCore/API/glib/JSCValue.cpp:
```diff
@@ -2145,3 +2145,61 @@ char* jsc_value_to_json(JSCValue* value, unsigned indent)
+        return nullptr;
+
+    const size_t argumentCount = 2;
+    JSValueRef arguments[argumentCount];
+    arguments[0] = resolveObj;
+    arguments[1] = rejectObj;
```
This broke the clang build when it landed today. Fix in: https://github.com/WebKit/WebKit/pull/38844
